bindings/rust-examples/key-logging/src/lib.rs

```diff
@@ -0,0 +1,168 @@
+        _conn: *mut s2n_connection,
+        logline: *mut u8,
+        len: usize,
+    ) -> ffi::c_int {
+        let handle = &mut *(ctx as *mut Self);
+        let logline = core::slice::from_raw_parts(logline, len);
+
+        // ignore any errors
+        let _ = handle.on_logline(logline);
+
+        0
+    }
+
+    fn on_logline(&mut self, logline: &[u8]) -> Option<()> {
+        let mut file = self.0.lock().ok()?;
+        file.write_all(logline).ok()?;
+        file.write_all(b"\n").ok()?;
+
+        // ensure keys are immediately written so tools can use them
+        file.flush().ok()?;
+
+        Some(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use s2n_tls::{config::Config, security::DEFAULT_TLS13};
+    use s2n_tls_tokio::{TlsAcceptor, TlsConnector};
+    use std::net::{Ipv4Addr, SocketAddr, SocketAddrV4};
+    use tokio::{
+        io::{AsyncReadExt, AsyncWriteExt},
+        net::{TcpListener, TcpStream},
+    };
+
+    use super::*;
+
+    /// NOTE: this CA is to be used for demonstration purposes only!
+    const CA: &[u8] = include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/../certs/ca-cert.pem"));
+    const CHAIN: &[u8] = include_bytes!(concat!(
+        env!("CARGO_MANIFEST_DIR"),
+        "/../certs/localhost-chain.pem"
+    ));
+    const KEY: &[u8] = include_bytes!(concat!(
+        env!("CARGO_MANIFEST_DIR"),
+        "/../certs/localhost-key.pem"
+    ));
+    const SERVER_MESSAGE: &[u8] = b"hello world";
+
+    async fn launch_server() -> anyhow::Result<SocketAddr> {
+        let mut config = Config::builder();
+        config.set_security_policy(&DEFAULT_TLS13)?;
+        config.load_pem(CHAIN, KEY)?;
+        config.set_max_blinding_delay(0)?;
+
+        let server = TlsAcceptor::new(config.build()?);
+
+        let listener = TcpListener::bind(&SocketAddrV4::new(Ipv4Addr::LOCALHOST, 0)).await?;
+        let addr = listener.local_addr()?;
+
+        tokio::spawn(async move {
+            loop {
+                let (stream, _peer_addr) = listener.accept().await.unwrap();
+
+                let server = server.clone();
+                tokio::spawn(async move {
+                    let mut tls = server.accept(stream).await?;
+                    tls.write_all(SERVER_MESSAGE).await?;
+                    tls.shutdown().await?;
+                    Ok::<(), anyhow::Error>(())
+                });
+            }
+        });
+
+        Ok(addr)
+    }
+
+    #[tokio::test]
+    async fn client_key_logging() -> anyhow::Result<()> {
+        const KEY_PATH: &str = "s2n_client.keys";
+
+        // do some tls stuff
+        {
+            let key_logger = TlsKeyLogger::from_path(KEY_PATH).unwrap();
+
+            let mut client_config = s2n_tls::config::Builder::new();
+            client_config.trust_pem(CA)?;
+            client_config.set_security_policy(&DEFAULT_TLS13)?;
+            unsafe {
+                // The s2n-tls API currently requires a raw C callback and a raw C "context"
```

Comment:
Can you add a note about how we want to improve this? Like ideally s2n-tls would own the Keylogger and none of our users would touch ffi stuff. I would link to the github issue.